react/leanShipping.js

```diff
@@ -343,47 +354,136 @@ function shouldShowFastest(cheapest, fastest) {
   const cheapestDetails = getStructuredOption(cheapest, CHEAPEST)
   const fastestDetails = getStructuredOption(fastest, FASTEST)
 
-  const hasCheapestLessPackages = cheapestDetails.packagesLength < fastestDetails.packagesLength
+  const hasCheapestLessPackages =
+    cheapestDetails.packagesLength < fastestDetails.packagesLength
   const fastestAndCheapestAreEqual = isEqual(cheapest, fastest)
 
   return fastest && !fastestAndCheapestAreEqual && !hasCheapestLessPackages
 }
 
-export function getLeanShippingOptions({
+function areSelectedSlasValid(logisticsInfo) {
+  const selectedSlas = logisticsInfo
+    .filter(logisticInfo => logisticInfo.selectedSla !== null)
+    .map(logisticInfo => logisticInfo.selectedSla)
+
+  return logisticsInfo.every(
+    logisticInfo =>
+      intersection(
+        logisticInfo.slas.map(sla => sla.id),
+        selectedSlas
+      ).length === 1
+  )
+}
+
+function setSelectedSlaIfValid({
+  logisticsInfo,
+  selectedSlas,
+  activeChannel,
+  isScheduledDeliveryActive,
+}) {
+  const newLogisticsInfo = setSelectedSla({
+    logisticsInfo,
+    selectedSlas,
+    activeChannel,
+    isScheduledDeliveryActive,
+  })
+  return areSelectedSlasValid(newLogisticsInfo) ? newLogisticsInfo : null
+}
+
+function getBestOption({
   logisticsInfo,
   activeChannel = DELIVERY,
   isScheduledDeliveryActive = false,
+  property,
 }) {
-  const arraysOfSlas = createArraysOfSlas(logisticsInfo)
-  const selectedSlas = {
-    cheapest: getMinSlaBy(arraysOfSlas, 'price'),
-    fastest: getMinSlaBy(arraysOfSlas, 'shippingEstimateInSeconds'),
+  const arrayOfSlas = createArrayOfSlas(logisticsInfo)
+
+  const currentSelectedSlas = arrayOfSlas.map(
+    (slas, index) =>
+      slas.filter(sla => sla.id === logisticsInfo[index].selectedSla)[0]
+  )
+  const bestIndividualSlas = getBestSlasBy(arrayOfSlas, property, false)
+  const bestConsistentSlas = getBestSlasBy(arrayOfSlas, property, true)
+
+  const currentLogisticsInfo = setSelectedSlaIfValid({
+    logisticsInfo,
+    selectedSlas: currentSelectedSlas,
+    activeChannel,
+    isScheduledDeliveryActive,
+  })
+
+  const bestIndividualLogisticsInfo = setSelectedSlaIfValid({
+    logisticsInfo,
+    selectedSlas: bestIndividualSlas,
+    activeChannel,
+    isScheduledDeliveryActive,
+  })
+
+  const bestConsistentLogisticsInfo = setSelectedSlaIfValid({
+    logisticsInfo,
+    selectedSlas: bestConsistentSlas,
+    activeChannel,
+    isScheduledDeliveryActive,
+  })
+
+  const validLogisticsInfoArray = [
+    currentLogisticsInfo,
+    bestIndividualLogisticsInfo,
+    bestConsistentLogisticsInfo,
+  ].filter(possibility => possibility)
+
+  const slasAvailableToEveryItem = intersection(
+    ...arrayOfSlas.map(slas => slas.map(sla => sla.id))
+  )
+
+  const bestLogisticsInfo = sortBy(validLogisticsInfoArray, [
```

Comment:
just to be sure (since you are getting a value straight from the function result), is this always guaranteed to be an array? (as opposed to null/undefined)

Reply:
also the same comment applies here regarding the usage of sort vs. getting just the best result

Reply:
`sortBy` is a function from lodash, [so they guarantee that its an array](https://lodash.com/docs/4.17.15#sortBy)